modules/table/src/test/java/org/apache/ignite/internal/table/distributed/storage/InternalTableEstimatedSizeTest.java

```diff
@@ -0,0 +1,471 @@
+import org.apache.ignite.internal.replicator.message.ReplicaRequest;
+import org.apache.ignite.internal.schema.SchemaRegistry;
+import org.apache.ignite.internal.storage.MvPartitionStorage;
+import org.apache.ignite.internal.storage.engine.MvTableStorage;
+import org.apache.ignite.internal.table.InternalTable;
+import org.apache.ignite.internal.table.distributed.StorageUpdateHandler;
+import org.apache.ignite.internal.table.distributed.index.IndexMetaStorage;
+import org.apache.ignite.internal.table.distributed.replicator.PartitionReplicaListener;
+import org.apache.ignite.internal.table.distributed.replicator.TransactionStateResolver;
+import org.apache.ignite.internal.table.distributed.schema.SchemaSyncService;
+import org.apache.ignite.internal.table.distributed.schema.ValidationSchemasSource;
+import org.apache.ignite.internal.testframework.BaseIgniteAbstractTest;
+import org.apache.ignite.internal.tx.HybridTimestampTracker;
+import org.apache.ignite.internal.tx.LockManager;
+import org.apache.ignite.internal.tx.TxManager;
+import org.apache.ignite.internal.tx.impl.RemotelyTriggeredResourceRegistry;
+import org.apache.ignite.internal.tx.impl.TransactionInflights;
+import org.apache.ignite.internal.tx.storage.state.TxStateStorage;
+import org.apache.ignite.internal.tx.storage.state.TxStateTableStorage;
+import org.apache.ignite.internal.util.Lazy;
+import org.apache.ignite.internal.util.PendingComparableValuesTracker;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.NetworkAddress;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInfo;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+/**
+ * Tests for distributed aspects of the {@link InternalTable#estimatedSize} method.
+ */
+@ExtendWith(MockitoExtension.class)
+@ExtendWith(ConfigurationExtension.class)
+public class InternalTableEstimatedSizeTest extends BaseIgniteAbstractTest {
+    private static final String TABLE_NAME = "TEST";
+
+    private static final int TABLE_ID = 1;
+
+    private static final int PARTITIONS_NUM = 3;
+
+    private static final List<TablePartitionId> TABLE_PARTITION_IDS = IntStream.range(0, PARTITIONS_NUM)
+            .mapToObj(i -> new TablePartitionId(TABLE_ID, i))
+            .collect(toList());
+
+    private ClusterNode node;
+
+    private InternalTableImpl table;
+
+    private MessagingService messagingService;
+
+    @Mock
+    private PlacementDriver placementDriver;
+
+    private final HybridClockImpl clock = new HybridClockImpl();
+
+    private final ComponentContext componentContext = new ComponentContext();
+
+    private final List<MvPartitionStorage> partitionStorages = new ArrayList<>(PARTITIONS_NUM);
+
+    private final List<IgniteComponent> components = new ArrayList<>();
+
+    @BeforeEach
+    void setUp(
```

Comment:
Great work first of all. However given test seems to be very fragile because it mainly tests mocks. Did you consider using real ignite nodes with some message blocking and transferPrimary/killing nodes? Such tests will be longer but more robust, WDYT?

Reply:
> Did you consider using real ignite nodes with some message blocking

I did, bu writing such tests and validating exact conditions that I want is a nightmare. I don't fully agree that it mainly tests mocks. My main goal was to see what exceptions are thrown by the `ReplicaService` in different conditions in order to check the exception types and these tests helped me achieve that